build.rs

```diff
@@ -1,10 +1,116 @@
-use std::env;
+use regex::Regex;
+use std::fmt::{Display, Formatter};
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use std::{env, fmt};
+
+#[derive(Clone, Debug)]
+pub struct Target {
+    pub architecture: String,
+    pub vendor: String,
+    pub system: Option<String>,
+    pub abi: Option<String>,
+}
+
+impl Target {
+    pub fn as_strs(&self) -> (&str, &str, Option<&str>, Option<&str>) {
+        (
+            self.architecture.as_str(),
+            self.vendor.as_str(),
+            self.system.as_deref(),
+            self.abi.as_deref(),
+        )
+    }
+}
+
+impl Display for Target {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "{}-{}", &self.architecture, &self.vendor)?;
+
+        if let Some(ref system) = self.system {
+            write!(f, "-{}", system)
+        } else {
+            Ok(())
+        }?;
+
+        if let Some(ref abi) = self.abi {
+            write!(f, "-{}", abi)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub fn ndk() -> String {
+    env::var("ANDROID_NDK").expect("ANDROID_NDK variable not set")
+}
+
+pub fn target_arch(arch: &str) -> &str {
+    match arch {
+        "armv7" => "arm",
+        "aarch64" => "arm64",
+        "i686" => "x86",
+        arch => arch,
+    }
+}
+
+fn host_tag() -> String {
```

Comment:
you only use this to interpolate into another String later; I would make this return `&'static str` and remove the `to_string`, saves you an allocation that you aren't using anyway.